### PR TITLE
Remove NameId from Core

### LINKED ISSUE #1491: Too many name ids generated in JuvixCore parser
The JuvixCore parser generates a fresh name id for each _occurrence_ of an identifier, not each definition. So e.g. `a + a` is parsed into `a@1 + a@2`.

This doesn't matter much because Core doesn't use the name ids anyway. But it would be nice to fix this.

### LINKED ISSUE #1564: Remove `Name` from Core in favour of separate name Text and Location fields
We don't use the nameIds in Core, so it's pointless and cumbersome to keep them around.

### PR DESCRIPTION
**Pull request by lukaszcz**
Fixes #1564
Fixes #1491
﻿

### CODE REVIEW

File: src/Juvix/Compiler/Core/Translation/FromInternal.hs [hunk 1]
```diff
@@ -374,15 +391,15 @@ goPatterns body ps = do
   vars <- asks (^. indexTableVars)
   varsNum <- asks (^. indexTableVarsNum)
   pats <- patterns
-  let bs :: [Binder]
-      bs = concatMap getPatternBinders pats
+  let bs :: [Name]
+      bs = concatMap getPatternVars (reverse ps)
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Translation/FromInternal.hs:395]: No, but `pats` is reversed with respect to `ps` (see definition below)

File: src/Juvix/Compiler/Core/Extra/Base.hs [hunk 2]
```diff
@@ -465,12 +465,13 @@ destruct = \case
               (values', tys') = second (map (^. childNode)) (splitAtExact numItems valuesTys')
               items' =
                 nonEmpty'
-                  [ LetItem (Binder name ty') (v' ^. childNode)
-                    | (v', ty', name) <-
-                        zip3Exact
+                  [ LetItem (Binder name loc ty') (v' ^. childNode)
+                    | (v', ty', name, loc) <-
+                        zip4
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Extra/Base.hs:470]: I would've kept it if zip4Exact existed. It goes only up to 3.

File: src/Juvix/Compiler/Core/Translation/FromInternal.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Translation/FromInternal.hs:395]: By the way, I don't understand what this reverse is for, but it doesn't seem to work without it. Are the patterns in the clauses in Internal in right-to-left order?

File: src/Juvix/Compiler/Core/Language/Stripped.hs [hunk 3]
```diff
@@ -15,23 +15,27 @@ import Juvix.Compiler.Core.Language.Stripped.Type
 {---------------------------------------------------------------------------------}
 
 data VarInfo = VarInfo
-  { _varInfoName :: Maybe Name,
+  { _varInfoName :: Text,
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Language/Stripped.hs:18]: I removed the checks for empty string in favour of always providing some name ("?" as default).

File: src/Juvix/Compiler/Core/Extra/Base.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (janmasrovira) [reply src/Juvix/Compiler/Core/Extra/Base.hs:470]: can we add it to prelude?
Or do something adhoc like:
```
(v', ty', (name, loc)) <- zip3Exact vs tys (zipExact names locs)

File: src/Juvix/Compiler/Core/Translation/FromInternal.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (janmasrovira) [reply src/Juvix/Compiler/Core/Translation/FromInternal.hs:395]: they are supposed to be in left-to-right order. I don't know what could cause this issue

**Review by janmasrovira — APPROVED** [review]
I've added a suggestion. The rest is fine

File: src/Juvix/Compiler/Core/Extra/Base.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Extra/Base.hs]: I added zip4Exact to Prelude.Base

**Pull request merged**